tasks/Task.cpp

```diff
@@ -1,109 +1,95 @@
     if (! TaskBase::configureHook())
         return false;
-
-    /* Opens serial port */
-    lamps.openSerial(_port.get(), _baud_rate.get());
-
-    lamps.setCmdLabel(_cmd_label.get());
-
-    led_list = _led_list.get();
-    light_level_all  = _light_level_all.get();
-
-
-    /* Sets initializing properties of the lamps */
-    for (int i = 0; i < led_list.size(); ++i)
-    {
-    	lamps.setPowerUpLightLevel(led_list[i].power_up_light_level, led_list[i].address);
-    	usleep(100000);
-    	lamps.setLightLevel(led_list[i].light_level, led_list[i].address);
-    	usleep(100000);
-    }
-
     return true;
 }
 bool Task::startHook()
 {
     if (! TaskBase::startHook())
         return false;
+
+    /* Set the initial light level */
+    lamps = _lamps.get();
+    for (unsigned int i = 0; i < lamps.size(); ++i)
+    	driver.setLightLevel(lamps[i].level, lamps[i].address);
+
     return true;
 }
 void Task::updateHook()
 {
     TaskBase::updateHook();
 
-    if (light_level_all != _light_level_all.get())
+    std::vector<Lamp> levels;
+    while (_levels.readNewest(levels) == RTT::NewData)
     {
-    	light_level_all = _light_level_all.get();
-
-    	/* This if-statement will set the light level of all lamps to the value
-    	 * specified in the property _light_level_all */
-    	if(light_level_all > 0)
-    	{
-    		lamps.setLightLevel(light_level_all);
-
-    		for(int i = 0; i < led_list.size(); i++)
-    			led_list[i].light_level = light_level_all;
-
-    		_led_list.set(led_list);
-    	}
-    }
-    else
-    {
-    	for(int i = 0; i < led_list.size(); i++)
-    	{
-    		/* If one of the lamps light level property has changed, the
-    		 * property _light_level_all will be set as negative, and the lamps'
-    		 * light level will be set individually */
-    		if(led_list[i].light_level != _led_list.get()[i].light_level)
-    		{
-    			light_level_all = -1;
-    			_light_level_all.set(light_level_all);
-    			led_list[i].light_level = _led_list.get()[i].light_level;
-
-    			lamps.setLightLevel(led_list[i].light_level, led_list[i].address);
-    		}
-    	}
+        for (unsigned int i = 0; i < levels.size(); ++i)
+        {
+            int address = levels[i].address;
+            if (!levels[i].name.empty())
+                address = resolveLightAddress(levels[i].name);
+
+            driver.setLightLevel(levels[i].level, address);
+        }
     }
+}
+
+void Task::processIO()
+{
+    // Nothing to do here (or is there ?)
+}
 
+uint8_t Task::resolveLightAddress(std::string const& name) const
+{
+    for (unsigned int i = 0; i < lamps.size(); ++i)
+        if (lamps[i].name == name)
+            return lamps[i].address;
+    throw std::invalid_argument("there is not ligth named " + name);
 }
+
 void Task::errorHook()
 {
     TaskBase::errorHook();
 }
 void Task::stopHook()
 {
+    /* Turn the lamps off */
+    for (unsigned int i = 0; i < lamps.size(); ++i)
+    	driver.setLightLevel(0, lamps[i].address);
```

Comment:
Hi @doudou,

according to the lamp's protocol, the light level cannot receive 0. I had a look at the driver and saw that it will cause an runtime_error.